crates/metal-ast/src/expr.rs

```diff
@@ -1,7 +1,16 @@
-use crate::Ident;
+use crate::{Ident, Path, Ty};
 
 /// An expression, also called a value.
 pub enum Expr<'src> {
     /// An identifier in place of an expression, like `arg` in `print(arg)`.
     Ident(Ident<'src>),
+    // TODO: represent this some other way
+    /// A number value, like `1` or `2.6`.
+    Number { ty: Ty<'src>, value: u64 },
```

Comment:
this is a detail I actually changed because of LLVM. `LLVMConstInt` requires a `ulonglong` or `u64`, I don't think negative values are supported. I will make an issue for this after this is merged and try to find a workaround.

Reply:
And `ty` just details what primitive integer type the number should be assigned to.

Reply:
> this is a detail I actually changed because of LLVM. `LLVMConstInt` requires a `ulonglong` or `u64`, I don't think negative values are supported. I will make an issue for this after this is merged and try to find a workaround.

`LLVMConstInt` has a third parameter, `SignExtend` (arguably confusing), which allows to turn a u64 into a negative u64. [Here](https://www.llvm.org/doxygen/group__LLVMCCoreValueConstantScalar.html#ga06540cbed28286f40b0f625bf05d0233) is the function in question. If we look at it's source/docs, we can see that it returns a `ConstantInt` by using it's `get` [method](https://www.llvm.org/doxygen/classllvm_1_1ConstantInt.html#a0e753b1d4f854db309aac9c6ed935159), whose third argument, with an actually sensible name this time, is `isSigned`. The way conversion from `SignExtend` to `isSigned` happens is `isSigned = SignExtend != 0`, therefore, to make a negative constant integer, one has to pass 1 (or any value other than 0) as the third argument to `LLVMConstInt`.